lib/mix/test/mix/tasks/compile.elixir_test.exs

```diff
@@ -236,6 +236,41 @@ defmodule Mix.Tasks.Compile.ElixirTest do
 
       assert Mix.Tasks.Compile.Elixir.run(["--elixirc-paths", "lib"])
       refute_received {:mix_shell, :info, ["Compiled lib/a.ex"]}
+
+      # Compiling just web and then just lib should recompile lib
+      # if both are modified
+      :timer.sleep(1000)
```

Comment:
`ensure_touched` defined in mix/test_helper.exs could be used. In the worst scenario it is 1s but on average it is much better.